numpy/core/numeric.py

```diff
@@ -48,6 +50,30 @@
     import __builtin__ as builtins
 
 
+# TODO: rewrite the __call__ in C?
+class UFuncWrapper(ufunc_wrapper):
+    ''' A decorator to wrap ufunc-like class methods and enable the
+        __array_ufunc__ protocol
+
+        Use as
+
+        class MyArray(ndarray):
+            @UFuncWrapper(2, 1) # nin, nout
+            def solve(self, other):
+                ...
+                return result
+    '''
+    def __call__(self, meth):
+        def wrap(obj, *args, **kwds):
+            r = self.check_override(*((self,) + args), **kwds)
```

Comment:
Shouldn't this be `*((obj,) + args)`? `check_override` would need to pass on the wrapped function, no?

Reply:
Sorry, I meant `meth` (which is the wrapped method, correct?) And then `obj` will be `self` of `ndarray`, which I think would not be passed on at all.

Reply:
AFAICT check_override does not need `meth` to call `PyUFunc_CheckOverride`, it only needs `self` to provide the `nin, nout` for argument normalization. The actual test for `__array_ufunc__` is in `PyUFunc_WithOverride` (from `src/private/ufunc_override.c`) which indeed does need `obj`, so that line should be
```
r = self.check_override(*((self, obj) + args), **kwds)
```
`obj` is then not different than args, and the code can be simplified to (after taking into account additional corrections)
```
    def __call__(self, meth):
        def wrap(*args, **kwds):
            # use status to indicate that one of args overrode the call
            (status, r) = self.check_override(*((self,) + args), **kwds)
            if status > 0:
                return r
            return meth(*args, **kwds)
```

Reply:
I do think `meth` needs to be passed on, and not the wrapper class as "ufunc", otherwise any class defining `__array_ufunc__` has no idea what the real operation is and thus cannot use it to act accordingly. E.g., for astropy's `Quantity`, I do need to know that the wrapped ufunc is `matmul`, since that means I know the units of the two arguments have to be multiplied.